src/utilities/bmi/mass_balance.cpp

```diff
@@ -0,0 +1,137 @@
+               << " ("+ctx.timestamp+")"
+               << " at feature id " << ctx.id <<std::endl
+               << "\tMass balance check failed for " << model->GetComponentName() << "\n\t" <<
+                INPUT_MASS_NAME << "(" << mass_in <<  ") - " <<
+                OUTPUT_MASS_NAME << " (" << mass_out <<  ") - " << 
+                STORED_MASS_NAME << " (" << mass_stored <<  ") - " << 
+                LEAKED_MASS_NAME << " (" << mass_leaked << ") = " << 
+                mass_balance << "\n\t" << "tolerance: " << tolerance << std::endl;
+            if(is_fatal)
+                throw MassBalanceError("mass_balance::error " + ss.str());
+            else{
+                std::cerr << "mass_balance::warning " + ss.str();
+            }
+                
+        }
+    }
+}
+
+void NgenMassBalance::check_support() {
+    if (model->is_model_initialized()) {
+        double mass_var;
+        std::vector<std::string> units;
+        units.reserve(4);
+        try{
+            for(const auto& name : 
+                {INPUT_MASS_NAME, OUTPUT_MASS_NAME, STORED_MASS_NAME, LEAKED_MASS_NAME}
+            ){
+                model->GetValue(name, &mass_var);
+                units.push_back( model->GetVarUnits(name) );
+            }
+            //Compare all other units to the first one (+1)
+            if( std::equal( units.begin()+1, units.end(), units.begin() ) ) {
+                this->is_supported = true;
+            }
+            else{
+                // It may be possible to do unit conversion and still do meaninful mass balance
+                // this could be added as an extended feature, but for now, I don't think this is
+                // worth the complexity.  It is, however, worth the sanity check performed here
+                // to ensure the units are consistent.
+                throw MassBalanceError( "mass balance variables have incosistent units, cannot perform mass balance" );
+            }
+        } catch (const std::exception &e) {
+            std::stringstream ss;
+            ss << "mass_balance::integration: Error getting mass balance values for module '" << model->GetComponentName() << "': " << e.what() << std::endl;
+            std::cout << ss.str();
+            this->is_supported= false;
+        }
+    } else {
+        throw std::runtime_error("Cannot check mass balance for uninitialized model.");
+    }
+}
+
+void NgenMassBalance::initialize(const geojson::PropertyMap& properties)
+{
+    //Ensure the model is capable of mass balance using the protocol
+    check_support();
+
+    //now check if the user has requested to use mass balance
+    auto protocol_it = properties.find(CONFIGURATION_KEY);
+    if ( is_supported && protocol_it != properties.end() ) {
+        geojson::PropertyMap mass_bal = protocol_it->second.get_values();
+
+        auto _it = mass_bal.find(TOLERANCE_KEY);
+        if( _it != mass_bal.end() ) tolerance = _it->second.as_real_number();
```

Comment:
Should we warn on the cases where this is `NaN` or `0` since that would effectively disable the protocol?

Reply:
I think 0 is a perfectly fine tolerance to configure, if one expects the subject model to be numerically exact. It's hard to build that way, but not impossible.

Reply:
thoughts here, @hellkite500?

Reply:
After some discussion, I think we do need to check/handle NaN here.